### PR TITLE
feat: custom Spectral rule to ensure content objects contain schema

### PR DESCRIPTION
**Pull request by barrett-schonefeld**
Purpose:
- Establish a pattern in the validator for adding custom spectral rules and custom spectral functions
- Ensure content objects contain a schema

Changes:
- Add custom spectral rule to ensure that content objects contain a schema
- Add directory to store custom spectral rules and functions
- Add code to retrieve and merge custom spectral rules

Tests:
- Add tests for each use case outlined in the spectral rule

### CODE REVIEW

**Review by mkistler — CHANGES_REQUESTED** [review]
This does seem to satisfy the requirement but I think it can be simplified while still achieving the same result.

File: src/spectral/rulesets/custom-rules/content-entry-contains-schema.yaml
Comment (barrett-schonefeld) [reply src/spectral/rulesets/custom-rules/content-entry-contains-schema.yaml:1]: In hindsight, I think it may not be necessary to break the rules into different files. In different conversations about adding custom Sepctral rules, people mentioned that we may not want to clutter the `.defaultsForSpectral.yaml` file with custom rules.

I would say breaking custom rules in separate files has the following benefits:

1. Keeps the `.defaultsForSpectral.yaml` small and digestable.
2. Test files related to each custom rule are easier to create/find/update because test files map to a corresponding file with the same name.
3. Small benefit: When we add custom functions, some rules will need to specify the custom function it uses in a `functions` field. If the custom rules are in separate files, each real will clearly define the list of custom functions it uses rather than have a global list of custom functions in `.defaultsForSpectral.yaml`. This is only a small benefit because the custom rule will explicitly say which custom function it uses in the rule definition.

Do you think the reasons above are enough to introduce a more complicated structure now, or should we consider breaking rules into separate files as it becomes apparent that it's necessary?

Comment (mkistler) [reply src/spectral/rulesets/custom-rules/content-entry-contains-schema.yaml:1]: My take:
1. This just trades complexity within one file for the complexity of having multiple files -- which includes the extra logic in this PR for merging the multiple files.
2. Test files can be named for the rule -- I think this provides equal correlation between the rules and test files.
3. Valid but I'm not sure sufficient rationale for separate files.

I'll add to this that Spectral follows the "all rules in a single file" structure:

https://github.com/stoplightio/spectral/blob/develop/src/rulesets/oas/index.json

So I still think the single file approach is superior.

Comment (barrett-schonefeld) [reply src/spectral/rulesets/custom-rules/content-entry-contains-schema.yaml:1]: Yes, I think you're right, and I will make the updates to move forward with a single file approach.

File: src/spectral/rulesets/.defaultsForSpectral.yaml
Comment (dpopp07) [inline src/spectral/rulesets/.defaultsForSpectral.yaml:102]: I'm not that well versed in Spectral syntax but didn't Mike make a comment earlier about how `components` didn't need to be checked? Do you know if this line is necessary?

**Review by dpopp07 — APPROVED** [review]
I think this looks good but I did have one question about the Spectral implementation to check on before merging (might want to wait for other reviews as well)

File: src/spectral/rulesets/.defaultsForSpectral.yaml
Comment (barrett-schonefeld) [reply src/spectral/rulesets/.defaultsForSpectral.yaml:102]: Yes, thank you for catching!

**Review by mkistler — CHANGES_REQUESTED** [review]
The code here looks good but another key part of creating our own custom rules is documenting them.  We probably need to add a new section to the README that describes each custom rule, like the `docs/reference/openapi-rules.md` in Spectral.

**Comment by barrett-schonefeld** [comment]
> The code here looks good but another key part of creating our own custom rules is documenting them. We probably need to add a new section to the README that describes each custom rule, like the `docs/reference/openapi-rules.md` in Spectral.

To keep the `README.md` from getting overcrowded with rule descriptions, I documented the rule in `docs/spectral-rules.md` and linked to the rule descriptions in the `README.MD`.

File: README.md
Comment (mkistler) [inline README.md:575]: I think this is in the wrong place.  From the user's point of view, the rules we are adding are "standard rules" of our "ibm:oas" ruleset.  So this belongs in the previous section.

File: docs/spectral-rules.md
Comment (mkistler) [inline docs/spectral-rules.md:3]:
```suggestion
This document outlines the custom Spectral rules implemented in the `ibm:oas` ruleset.
```

**Review by mkistler — CHANGES_REQUESTED** [review]
Almost done, but we need to document this from the point of view of the user.

**Review by mkistler — APPROVED** [review]
Looks good. 👍

**Pull request merged**
